File: solutions/solver.cpp
#include <iostream>
#include <string>
#include <vector>
#include <memory>

using namespace std;

enum TermType { VAR, ABS, APP, NUM, BINOP };
enum BinOpKind { ADD, SUB, MUL };

struct Term {
    TermType type;
    string var_name;
    shared_ptr<Term> left, right, body;
    long long num_val;
    BinOpKind op;
    
    Term(TermType t) : type(t), num_val(0), op(ADD) {}
};

using TermPtr = shared_ptr<Term>;

inline vector<string> tokenize(const string& input) {
    vector<string> result;
    int i = 0, n = input.length();
    
    while (i < n) {
        char c = input[i];
        
        if (c <= ' ') {
            i++;
            continue;
        }
        
        if (c == '(' || c == ')' || c == '\\' || c == '.' || 
            c == '+' || c == '-' || c == '*') {
            result.push_back(string(1, c));
            i++;
        } else if (c >= '0' && c <= '9') {
            string num;
            while (i < n && input[i] >= '0' && input[i] <= '9') {
                num += input[i++];
            }
            result.push_back(num);
        } else {
            string var;
            while (i < n && ((input[i] >= 'a' && input[i] <= 'z') || 
                            (input[i] >= 'A' && input[i] <= 'Z') ||
                            (input[i] >= '0' && input[i] <= '9') || input[i] == '_')) {
                var += input[i++];
            }
            result.push_back(var);
        }
    }
    
    return result;
}

inline bool is_binop(const string& token) {
    return token == "+" || token == "-" || token == "*";
}

TermPtr parse_expr(const vector<string>& tokens, int& pos);

inline TermPtr parse_atom(const vector<string>& tokens, int& pos) {
    if (tokens[pos] == "(") {
        pos++;
        auto term = parse_expr(tokens, pos);
        pos++; // skip ')'
        return term;
    } else if (tokens[pos] == "\\") {
        pos++;
        string var = tokens[pos++];
        pos++; // skip '.'
        
        auto body = parse_expr(tokens, pos);
        auto abs_term = make_shared<Term>(ABS);
        abs_term->var_name = var;
        abs_term->body = body;
        return abs_term;
    } else if (tokens[pos][0] >= '0' && tokens[pos][0] <= '9') {
        auto num_term = make_shared<Term>(NUM);
        num_term->num_val = stoll(tokens[pos++]);
        return num_term;
    } else {
        auto var_term = make_shared<Term>(VAR);
        var_term->var_name = tokens[pos++];
        return var_term;
    }
}

inline TermPtr parse_term(const vector<string>& tokens, int& pos) {
    auto term = parse_atom(tokens, pos);
    
    while (pos < tokens.size() && 
           tokens[pos] != ")" && 
           tokens[pos] != "." &&
           !is_binop(tokens[pos])) {
        auto arg = parse_atom(tokens, pos);
        
        auto app_term = make_shared<Term>(APP);
        app_term->left = term;
        app_term->right = arg;
        term = app_term;
    }
    
    return term;
}

inline TermPtr parse_mul(const vector<string>& tokens, int& pos) {
    auto left = parse_term(tokens, pos);
    
    while (pos < tokens.size() && tokens[pos] == "*") {
        pos++;
        auto right = parse_term(tokens, pos);
        
        auto binop_term = make_shared<Term>(BINOP);
        binop_term->left = left;
        binop_term->right = right;
        binop_term->op = MUL;
        left = binop_term;
    }
    
    return left;
}

inline TermPtr parse_add(const vector<string>& tokens, int& pos) {
    auto left = parse_mul(tokens, pos);
    
    while (pos < tokens.size()) {
        BinOpKind op;
        if (tokens[pos] == "+") {
            op = ADD;
        } else if (tokens[pos] == "-") {
            op = SUB;
        } else {
            break;
        }
        pos++;
        
        auto right = parse_mul(tokens, pos);
        
        auto binop_term = make_shared<Term>(BINOP);
        binop_term->left = left;
        binop_term->right = right;
        binop_term->op = op;
        left = binop_term;
    }
    
    return left;
}

inline TermPtr parse_expr(const vector<string>& tokens, int& pos) {
    return parse_add(tokens, pos);
}

inline TermPtr parse(const string& input) {
    auto tokens = tokenize(input);
    int pos = 0;
    return parse_expr(tokens, pos);
}

TermPtr substitute(TermPtr term, const string& var, TermPtr replacement) {
    switch (term->type) {
        case VAR:
            return (term->var_name == var) ? replacement : term;
            
        case NUM:
            return term;
            
        case ABS:
            if (term->var_name == var) {
                return term;
            } else {
                auto new_abs = make_shared<Term>(ABS);
                new_abs->var_name = term->var_name;
                new_abs->body = substitute(term->body, var, replacement);
                return new_abs;
            }
            
        case APP: {
            auto new_app = make_shared<Term>(APP);
            new_app->left = substitute(term->left, var, replacement);
            new_app->right = substitute(term->right, var, replacement);
            return new_app;
        }
        
        case BINOP: {
            auto new_binop = make_shared<Term>(BINOP);
            new_binop->left = substitute(term->left, var, replacement);
            new_binop->right = substitute(term->right, var, replacement);
            new_binop->op = term->op;
            return new_binop;
        }
    }
    return term;
}

TermPtr evaluate(TermPtr term) {
    switch (term->type) {
        case VAR:
        case NUM:
            return term;
            
        case ABS: {
            auto new_abs = make_shared<Term>(ABS);
            new_abs->var_name = term->var_name;
            new_abs->body = evaluate(term->body);
            return new_abs;
        }
        
        case APP: {
            auto func = evaluate(term->left);
            auto arg = evaluate(term->right);
            
            if (func->type == ABS) {
                auto substituted = substitute(func->body, func->var_name, arg);
                return evaluate(substituted);
            } else {
                auto new_app = make_shared<Term>(APP);
                new_app->left = func;
                new_app->right = arg;
                return new_app;
            }
        }
        
        case BINOP: {
            auto left = evaluate(term->left);
            auto right = evaluate(term->right);
            
            if (left->type == NUM && right->type == NUM) {
                long long result;
                switch (term->op) {
                    case ADD: result = left->num_val + right->num_val; break;
                    case SUB: result = left->num_val - right->num_val; break;
                    case MUL: result = left->num_val * right->num_val; break;
                }
                auto num_term = make_shared<Term>(NUM);
                num_term->num_val = result;
                return num_term;
            } else {
                auto new_binop = make_shared<Term>(BINOP);
                new_binop->left = left;
                new_binop->right = right;
                new_binop->op = term->op;
                return new_binop;
            }
        }
    }
    return term;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int n;
    cin >> n;
    cin.ignore();
    
    for (int i = 0; i < n; i++) {
        string line;
        getline(cin, line);
        
        auto term = parse(line);
        auto result = evaluate(term);
        cout << result->num_val << '\n';
    }
    
    return 0;
}

File: solutions/solver.py
#!/usr/bin/env python3
"""
Lambda calculus evaluator for competitive programming.
Supports parsing and evaluating lambda expressions with arithmetic operations.
"""

import re
import sys
from typing import List, Union, Optional
from enum import Enum
from dataclasses import dataclass

# Increase recursion limit for deep lambda expressions
sys.setrecursionlimit(20000)


class TermType(Enum):
    VAR = "variable"
    ABS = "abstraction"
    APP = "application"
    NUM = "number"
    BINOP = "binary_operation"


class BinOpKind(Enum):
    ADD = "+"
    SUB = "-"
    MUL = "*"


@dataclass
class Term:
    """Represents a lambda calculus term."""
    term_type: TermType
    var_name: Optional[str] = None
    left: Optional['Term'] = None
    right: Optional['Term'] = None
    body: Optional['Term'] = None
    num_val: Optional[int] = None
    op: Optional[BinOpKind] = None


class LambdaCalculusSolver:
    """Lambda calculus parser and evaluator."""
    
    def __init__(self):
        pass
    
    def tokenize(self, input_str: str) -> List[str]:
        """Tokenize the input string into a list of tokens."""
        tokens = []
        i = 0
        n = len(input_str)
        
        while i < n:
            c = input_str[i]
            
            # Skip whitespace
            if c.isspace():
                i += 1
                continue
            
            # Single character tokens
            if c in "()\\+*-.":
                tokens.append(c)
                i += 1
            # Numbers
            elif c.isdigit():
                num = ""
                while i < n and input_str[i].isdigit():
                    num += input_str[i]
                    i += 1
                tokens.append(num)
            # Variables/identifiers
            elif c.isalpha() or c == '_':
                var = ""
                while i < n and (input_str[i].isalnum() or input_str[i] == '_'):
                    var += input_str[i]
                    i += 1
                tokens.append(var)
            else:
                i += 1  # Skip unknown characters
        
        return tokens
    
    def is_binop(self, token: str) -> bool:
        """Check if token is a binary operator."""
        return token in ["+", "-", "*"]
    
    def parse_atom(self, tokens: List[str], pos: List[int]) -> Term:
        """Parse an atomic expression."""
        if pos[0] >= len(tokens):
            raise ValueError("Unexpected end of input")
        
        token = tokens[pos[0]]
        
        if token == "(":
            pos[0] += 1  # skip '('
            term = self.parse_expr(tokens, pos)
            if pos[0] >= len(tokens) or tokens[pos[0]] != ")":
                raise ValueError("Expected closing parenthesis")
            pos[0] += 1  # skip ')'
            return term
        elif token == "\\":
            pos[0] += 1  # skip '\'
            if pos[0] >= len(tokens):
                raise ValueError("Expected variable after \\")
            var = tokens[pos[0]]
            pos[0] += 1
            if pos[0] >= len(tokens) or tokens[pos[0]] != ".":
                raise ValueError("Expected '.' after variable")
            pos[0] += 1  # skip '.'
            body = self.parse_expr(tokens, pos)
            return Term(TermType.ABS, var_name=var, body=body)
        elif token.isdigit():
            pos[0] += 1
            return Term(TermType.NUM, num_val=int(token))
        else:
            # Variable
            pos[0] += 1
            return Term(TermType.VAR, var_name=token)
    
    def parse_term(self, tokens: List[str], pos: List[int]) -> Term:
        """Parse a term with function applications."""
        term = self.parse_atom(tokens, pos)
        
        # Handle function applications (left-associative)
        while (pos[0] < len(tokens) and 
               tokens[pos[0]] != ")" and 
               tokens[pos[0]] != "." and
               not self.is_binop(tokens[pos[0]])):
            arg = self.parse_atom(tokens, pos)
            term = Term(TermType.APP, left=term, right=arg)
        
        return term
    
    def parse_mul(self, tokens: List[str], pos: List[int]) -> Term:
        """Parse multiplication (higher precedence)."""
        left = self.parse_term(tokens, pos)
        
        while pos[0] < len(tokens) and tokens[pos[0]] == "*":
            pos[0] += 1  # consume '*'
            right = self.parse_term(tokens, pos)
            left = Term(TermType.BINOP, left=left, right=right, op=BinOpKind.MUL)
        
        return left
    
    def parse_add(self, tokens: List[str], pos: List[int]) -> Term:
        """Parse addition and subtraction (lower precedence)."""
        left = self.parse_mul(tokens, pos)
        
        while pos[0] < len(tokens):
            token = tokens[pos[0]]
            if token == "+":
                op = BinOpKind.ADD
            elif token == "-":
                op = BinOpKind.SUB
            else:
                break
            
            pos[0] += 1
            right = self.parse_mul(tokens, pos)
            left = Term(TermType.BINOP, left=left, right=right, op=op)
        
        return left
    
    def parse_expr(self, tokens: List[str], pos: List[int]) -> Term:
        """Parse a complete expression."""
        return self.parse_add(tokens, pos)
    
    def parse(self, input_str: str) -> Term:
        """Parse a lambda expression from string."""
        tokens = self.tokenize(input_str)
        pos = [0]  # Use list to make it mutable in nested calls
        return self.parse_expr(tokens, pos)
    
    def substitute(self, term: Term, var: str, replacement: Term) -> Term:
        """Substitute variable with replacement term."""
        if term.term_type == TermType.VAR:
            if term.var_name == var:
                return replacement
            else:
                return term
        elif term.term_type == TermType.NUM:
            return term
        elif term.term_type == TermType.ABS:
            if term.var_name == var:
                # Variable is shadowed
                return term
            else:
                new_body = self.substitute(term.body, var, replacement)
                return Term(TermType.ABS, var_name=term.var_name, body=new_body)
        elif term.term_type == TermType.APP:
            new_left = self.substitute(term.left, var, replacement)
            new_right = self.substitute(term.right, var, replacement)
            return Term(TermType.APP, left=new_left, right=new_right)
        elif term.term_type == TermType.BINOP:
            new_left = self.substitute(term.left, var, replacement)
            new_right = self.substitute(term.right, var, replacement)
            return Term(TermType.BINOP, left=new_left, right=new_right, op=term.op)
        
        return term
    
    def evaluate(self, term: Term) -> Term:
        """Evaluate a lambda expression using beta reduction."""
        if term.term_type in [TermType.VAR, TermType.NUM]:
            return term
        elif term.term_type == TermType.ABS:
            new_body = self.evaluate(term.body)
            return Term(TermType.ABS, var_name=term.var_name, body=new_body)
        elif term.term_type == TermType.APP:
            func = self.evaluate(term.left)
            arg = self.evaluate(term.right)
            
            if func.term_type == TermType.ABS:
                # Beta reduction: (λx.M) N → M[x := N]
                substituted = self.substitute(func.body, func.var_name, arg)
                return self.evaluate(substituted)
            else:
                return Term(TermType.APP, left=func, right=arg)
        elif term.term_type == TermType.BINOP:
            left = self.evaluate(term.left)
            right = self.evaluate(term.right)
            
            if left.term_type == TermType.NUM and right.term_type == TermType.NUM:
                if term.op == BinOpKind.ADD:
                    result = left.num_val + right.num_val
                elif term.op == BinOpKind.SUB:
                    result = left.num_val - right.num_val
                elif term.op == BinOpKind.MUL:
                    result = left.num_val * right.num_val
                else:
                    raise ValueError(f"Unknown operator: {term.op}")
                
                return Term(TermType.NUM, num_val=result)
            else:
                return Term(TermType.BINOP, left=left, right=right, op=term.op)
        
        return term


def main():
    """Main function to read input and process lambda expressions."""
    solver = LambdaCalculusSolver()
    
    n = int(input().strip())
    
    for _ in range(n):
        line = input().strip()
        try:
            term = solver.parse(line)
            result = solver.evaluate(term)
            print(result.num_val)
        except Exception as e:
            print(f"Error: {e}")


if __name__ == "__main__":
    main()